debezium-embedded/src/main/java/io/debezium/embedded/AsyncEmbeddedEngine.java

```diff
@@ -0,0 +1,1266 @@
+
+        // Instantiate remaining required objects.
+        try {
+            this.offsetCommitPolicy = offsetCommitPolicy == null
+                    ? Instantiator.getInstanceWithProperties(this.config.getString(AsyncEngineConfig.OFFSET_COMMIT_POLICY), config)
+                    : offsetCommitPolicy;
+            offsetKeyConverter = Instantiator.getInstance(JsonConverter.class.getName());
+            offsetValueConverter = Instantiator.getInstance(JsonConverter.class.getName());
+            transformations = new Transformations(Configuration.from(config));
+
+            final Class<? extends SourceConnector> connectorClass = (Class<SourceConnector>) this.classLoader
+                    .loadClass(this.config.getString(AsyncEngineConfig.CONNECTOR_CLASS));
+            final SourceConnector connectConnector = connectorClass.getDeclaredConstructor().newInstance();
+            this.connector = new EngineSourceConnector(connectConnector);
+        }
+        catch (Throwable t) {
+            this.completionCallback.handle(false, "Failed to instantiate required class", t);
+            throw new DebeziumException(t);
+        }
+
+        // Disable schema for default JSON converters used for offset store.
+        Map<String, String> internalConverterConfig = Collections.singletonMap(JsonConverterConfig.SCHEMAS_ENABLE_CONFIG, "false");
+        offsetKeyConverter.configure(internalConverterConfig, true);
+        offsetValueConverter.configure(internalConverterConfig, false);
+    }
+
+    @Override
+    public void run() {
+        Throwable exitError = null;
+        try {
+            LOGGER.debug("Initializing connector and starting it.");
+            setEngineState(State.CREATING, State.INITIALIZING);
+            connector.connectConnector().start(initializeConnector());
+            LOGGER.debug("Calling connector callback after connector has started.");
+            connectorCallback.ifPresent(DebeziumEngine.ConnectorCallback::connectorStarted);
+
+            LOGGER.debug("Creating source tasks.");
+            setEngineState(State.INITIALIZING, State.CREATING_TASKS);
+            createSourceTasks(connector, tasks);
+
+            LOGGER.debug("Starting source tasks.");
+            setEngineState(State.CREATING_TASKS, State.STARING_TASKS);
+            startSourceTasks(tasks);
+
+            LOGGER.debug("Starting tasks polling.");
+            setEngineState(State.STARING_TASKS, State.POLLING_TASKS);
+            runTasksPolling(tasks);
+            // Tasks run infinite polling loop until close() is called or exception is thrown.
+        }
+        catch (Throwable t) {
+            exitError = t;
+            LOGGER.error("Engine has failed with ", exitError);
+
+            final State stateBeforeStop = getEngineState();
+            // Skip shutting down the engine if it's already being stopped.
+            if (State.STOPPING.compareTo(stateBeforeStop) > 0) {
```

Comment:
Is this safe? Maybe the enum can provide method like `canBeStopped` and called here.